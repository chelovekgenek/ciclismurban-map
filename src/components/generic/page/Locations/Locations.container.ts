import { withRouter } from "react-router"

import { Locations } from "./Locations"

export const LocationsContainer = withRouter(Locations)
