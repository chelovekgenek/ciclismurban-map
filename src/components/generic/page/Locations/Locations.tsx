import React, { useCallback, useMemo, useEffect } from "react"
import { RouteComponentProps } from "react-router"

import { Button, Table, Icon, Popconfirm } from "components/generic/ui"
import { App } from "components/generic/layout"
import { getGMapLink } from "helpers"
import { LocationModel } from "models/location"

interface IProps extends RouteComponentProps {
  title: string
  links: {
    create: string
    getInfo: (uuid: string) => string
  }
  locations: LocationModel[]
  fetching: boolean
  getLocations: () => {}
  deleteLocation: (uuid: string) => {}
}

export const Locations: React.FC<IProps> = ({
  history,
  title,
  links,
  locations,
  fetching,
  getLocations,
  deleteLocation,
}) => {
  useEffect(() => {
    getLocations()
  }, [getLocations])
  const handleCreate = useCallback(() => history.push(links.create), [history, links.create])
  const handleRowKey = useCallback((record: typeof locations[number]) => record.uuid, [locations])
  const columns = useMemo(
    () => [
      {
        title: "Название",
        dataIndex: "title",
      },
      {
        render: ({ point }: typeof locations[number]) => (
          <a target="_blank" rel="noopener noreferrer" href={getGMapLink(point)}>
            Координаты
          </a>
        ),
      },
      {
        render: ({ uuid }: typeof locations[number]) => (
          <Button.Group>
            <Button onClick={() => history.push(links.getInfo(uuid))}>
              <Icon type="edit" />
            </Button>
            <Popconfirm
              title="Вы уверены?"
              cancelText="Отменить"
              okText="Подтвердить"
              onConfirm={() => deleteLocation(uuid)}
            >
              <Button type="danger" disabled={fetching}>
                <Icon type="delete" />
              </Button>
            </Popconfirm>
          </Button.Group>
        ),
      },
    ],
    [locations, fetching, deleteLocation, links, history],
  )
  return (
    <App
      content={{
        useHeader: true,
        useLayout: true,
        title,
        actions: [
          <Button key="1" type="primary" onClick={handleCreate}>
            Создать
          </Button>,
        ],
      }}
    >
      <Table columns={columns} dataSource={locations} rowKey={handleRowKey} />
    </App>
  )
}
