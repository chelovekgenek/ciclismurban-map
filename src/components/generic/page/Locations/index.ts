export { LocationsContainer as Locations } from "./Locations.container"
