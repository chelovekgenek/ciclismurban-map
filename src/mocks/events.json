[
  {
    "startedAt": "2019-08-09T21:32:23.635Z",
    "uuid": "5d4db52b-4420-4e24-b41f-b02ccd8a6444",
    "title": "new",
    "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
    "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/cb858bf0-629e-49b5-86e4-17a0f4276456.jpg",
    "point": {
      "lat": 47.025838128285315,
      "lng": 28.834142687866233
    },
    "createdAt": "2019-08-09T21:36:10.967Z",
    "updatedAt": "2019-10-23T14:21:06.708Z"
  },
  {
    "startedAt": "2019-08-13T11:53:17.860Z",
    "uuid": "606601b8-d34c-481d-91a0-78c96798e8ef",
    "title": "the most new events",
    "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
    "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/cefe4cc3-e4d0-4246-b276-8416bbc2fec7.png",
    "point": {
      "lat": 47.035140101590976,
      "lng": 28.84796142871096
    },
    "createdAt": "2019-08-13T11:53:51.740Z",
    "updatedAt": "2019-08-13T11:53:51.740Z"
  },
  {
    "startedAt": "2019-10-23T14:21:15.016Z",
    "uuid": "e1f2b34e-ca13-4140-afb4-c4c2b1f6462b",
    "title": "privet",
    "description": "medved",
    "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/62408fd2-e36d-45bf-8218-18ca2ce6f359.jpg",
    "point": {
      "lat": 47.019518882203556,
      "lng": 28.80779266650393
    },
    "createdAt": "2019-10-23T14:22:05.604Z",
    "updatedAt": "2019-10-23T14:22:05.604Z"
  },
  {
    "startedAt": "2019-10-23T14:22:23.267Z",
    "uuid": "c30ea5e5-2915-41d6-b57c-8a8350492315",
    "title": "event",
    "description": "event event event",
    "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/90ce434a-423e-47cf-8cbb-31169e7a7799.jpg",
    "point": {
      "lat": 47.02864644200455,
      "lng": 28.838520052978538
    },
    "createdAt": "2019-10-23T14:22:47.357Z",
    "updatedAt": "2019-10-23T14:22:47.357Z"
  }
]
