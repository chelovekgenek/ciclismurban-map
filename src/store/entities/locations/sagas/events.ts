import { takeLatest, put, call } from "redux-saga/effects"
import { AxiosResponse } from "axios"
import { extend } from "lodash-es"

import { history } from "store/history"
import { EventModel } from "models/location"

import {
  EventsGetTypes,
  EventsCreateTypes,
  EventsDeleteTypes,
  EventsGetActions,
  EventsCreateActions,
  EventsDeleteActions,
  EventsUpdateActions,
  EventsUpdateTypes,
  SelectedActions,
} from "../actions"
import { createEvent, uploadFile, getEvents, deleteEvent, updateEvent } from "../api"

function* handleGet() {
  try {
    const { data }: AxiosResponse<EventModel[]> = yield call(getEvents)
    yield put(EventsGetActions.success(data))
  } catch (e) {
    yield put(EventsGetActions.failure(e))
  }
}

function* handleCreate({ payload: { image, ...payload } }: ReturnType<typeof EventsCreateActions.request>) {
  try {
    const { data: link }: AxiosResponse<string> = yield call(uploadFile, image!)
    const { data }: AxiosResponse<EventModel> = yield call(createEvent, { ...payload, image: link })
    yield put(EventsCreateActions.success(data))
    yield call(history.replace, { pathname: "/events" })
  } catch (e) {
    yield put(EventsCreateActions.failure(e))
  }
}

function* handleUpdate({ payload: { uuid, payload } }: ReturnType<typeof EventsUpdateActions.request>) {
  try {
    const reqPayload = extend(payload, {
      image: payload.image ? (yield call(uploadFile, payload.image)).data : undefined,
    })
    const { data }: AxiosResponse<EventModel> = yield call(updateEvent, uuid, reqPayload)
    yield put(EventsUpdateActions.success(data))
    yield put(SelectedActions.set(data))
  } catch (e) {
    yield put(EventsUpdateActions.failure(e))
  }
}

function* handleDelete({ payload }: ReturnType<typeof EventsDeleteActions.request>) {
  try {
    const { data }: AxiosResponse<string> = yield call(deleteEvent, payload)
    yield put(EventsDeleteActions.success(data))
  } catch (e) {
    yield put(EventsDeleteActions.failure(e))
  }
}

export function* watcher() {
  yield takeLatest(EventsGetTypes.REQUEST, handleGet)
  yield takeLatest(EventsCreateTypes.REQUEST, handleCreate)
  yield takeLatest(EventsUpdateTypes.REQUEST, handleUpdate)
  yield takeLatest(EventsDeleteTypes.REQUEST, handleDelete)
}
