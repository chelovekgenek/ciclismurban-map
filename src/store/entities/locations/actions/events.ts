import { payload, action } from "ts-action"

import { IState as IEventsState } from "../reducers/events"
import { LocationModel } from "models/location"

export enum EventsTypes {
  GET__REQUEST = "LOCATIONS__EVENTS__GET__REQUEST",
  GET__SUCCESS = "LOCATIONS__EVENTS__GET__SUCCESS",
  GET__FAILURE = "LOCATIONS__EVENTS__GET__FAILURE",
  CREATE__REQUEST = "LOCATIONS__EVENTS__CREATE__REQUEST",
  CREATE__SUCCESS = "LOCATIONS__EVENTS__CREATE__SUCCESS",
  CREATE__FAILURE = "LOCATIONS__EVENTS__CREATE__FAILURE",
}

type TRequiredEventsState = Required<IEventsState>

export const EventsActions = {
  requestGet: action(EventsTypes.GET__REQUEST),
  successGet: action(EventsTypes.GET__SUCCESS, payload<TRequiredEventsState["data"]>()),
  failureGet: action(EventsTypes.GET__FAILURE, payload<TRequiredEventsState["error"]>()),

  requestCreate: action(EventsTypes.GET__REQUEST, payload<LocationModel>()),
  successCreate: action(EventsTypes.GET__SUCCESS, payload<TRequiredEventsState["data"][number]>()),
  failureCreate: action(EventsTypes.GET__FAILURE, payload<TRequiredEventsState["error"]>()),
}
