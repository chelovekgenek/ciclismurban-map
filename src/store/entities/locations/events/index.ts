import * as Actions from "./events.actions"
import * as Selectors from "./events.selectors"

export { Actions, Selectors }
