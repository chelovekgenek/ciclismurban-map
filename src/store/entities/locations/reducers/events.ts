import { reducer, on, union } from "ts-action"

import { EventModel } from "models/location"

import { EventsGetActions, EventsCreateActions, EventsDeleteActions, EventsUpdateActions } from "../actions"

export interface IState {
  data: EventModel[]
  fetching: boolean
  error?: Error
}

const initialState: IState = {
  data: [],
  fetching: false,
}

export default reducer(
  initialState,
  on(
    ...union(
      EventsGetActions.request,
      EventsCreateActions.request,
      EventsUpdateActions.request,
      EventsDeleteActions.request,
    ),
    state => ({
      ...state,
      fetching: true,
    }),
  ),
  on(
    ...union(
      EventsGetActions.failure,
      EventsCreateActions.failure,
      EventsUpdateActions.failure,
      EventsDeleteActions.failure,
    ),
    (state, { payload }) => ({
      ...state,
      fetching: false,
      error: payload,
    }),
  ),
  on(EventsGetActions.success, (state, { payload }) => ({ ...state, fetching: false, data: payload })),
  on(EventsCreateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.concat(payload),
  })),
  on(EventsUpdateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.map(item => (item.uuid === payload.uuid ? payload : item)),
  })),
  on(EventsDeleteActions.success, (state, { payload }) => ({
    ...state,
    data: state.data.filter(item => item.uuid !== payload),
    fetching: false,
  })),
)
