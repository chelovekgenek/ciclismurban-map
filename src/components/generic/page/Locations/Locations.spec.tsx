import React from "react"
import { shallow } from "enzyme"

import events from "mocks/events.json"

import { Locations } from "./Locations"

describe("generic/page/Locations", () => {
  describe("pages/EventForm", () => {
    const render = (props: Partial<React.ComponentProps<typeof Locations>> = {}) =>
      shallow(
        <Locations
          title=""
          links={{ create: "", getInfo: (v: string) => v }}
          locations={events}
          fetching={false}
          getLocations={jest.fn() as any}
          deleteLocation={jest.fn() as any}
          create={jest.fn() as any}
          match={{ params: {} } as any}
          history={{} as any}
          location={{} as any}
          {...props}
        />,
      )
    it("should match snapshot", () => {
      expect(render()).toMatchSnapshot()
    })
  })
})
