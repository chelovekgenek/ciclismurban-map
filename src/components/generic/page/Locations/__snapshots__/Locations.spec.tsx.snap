// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic/page/Locations pages/EventForm should match snapshot 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="primary"
        >
          Создать
        </Button>,
      ],
      "title": "",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Table
    bordered={false}
    childrenColumnName="children"
    className=""
    columns={
      Array [
        Object {
          "dataIndex": "title",
          "title": "Название",
        },
        Object {
          "render": [Function],
        },
        Object {
          "render": [Function],
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "createdAt": "2019-08-09T21:36:10.967Z",
          "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
          "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/cb858bf0-629e-49b5-86e4-17a0f4276456.jpg",
          "point": Object {
            "lat": 47.025838128285315,
            "lng": 28.834142687866233,
          },
          "startedAt": "2019-08-09T21:32:23.635Z",
          "title": "new",
          "updatedAt": "2019-10-23T14:21:06.708Z",
          "uuid": "5d4db52b-4420-4e24-b41f-b02ccd8a6444",
        },
        Object {
          "createdAt": "2019-08-13T11:53:51.740Z",
          "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
          "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/cefe4cc3-e4d0-4246-b276-8416bbc2fec7.png",
          "point": Object {
            "lat": 47.035140101590976,
            "lng": 28.84796142871096,
          },
          "startedAt": "2019-08-13T11:53:17.860Z",
          "title": "the most new events",
          "updatedAt": "2019-08-13T11:53:51.740Z",
          "uuid": "606601b8-d34c-481d-91a0-78c96798e8ef",
        },
        Object {
          "createdAt": "2019-10-23T14:22:05.604Z",
          "description": "medved",
          "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/62408fd2-e36d-45bf-8218-18ca2ce6f359.jpg",
          "point": Object {
            "lat": 47.019518882203556,
            "lng": 28.80779266650393,
          },
          "startedAt": "2019-10-23T14:21:15.016Z",
          "title": "privet",
          "updatedAt": "2019-10-23T14:22:05.604Z",
          "uuid": "e1f2b34e-ca13-4140-afb4-c4c2b1f6462b",
        },
        Object {
          "createdAt": "2019-10-23T14:22:47.357Z",
          "description": "event event event",
          "image": "https://ciclismurban.s3.eu-central-1.amazonaws.com/files/90ce434a-423e-47cf-8cbb-31169e7a7799.jpg",
          "point": Object {
            "lat": 47.02864644200455,
            "lng": 28.838520052978538,
          },
          "startedAt": "2019-10-23T14:22:23.267Z",
          "title": "event",
          "updatedAt": "2019-10-23T14:22:47.357Z",
          "uuid": "c30ea5e5-2915-41d6-b57c-8a8350492315",
        },
      ]
    }
    indentSize={20}
    loading={false}
    locale={Object {}}
    rowKey={[Function]}
    showHeader={true}
    size="default"
    sortDirections={
      Array [
        "ascend",
        "descend",
      ]
    }
    useFixedHeader={false}
  />
</App>
`;
